hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/multitenant/TestRangerBGSyncService.java

```diff
@@ -0,0 +1,688 @@
+      LOG.info("Creating BucketAccess policy in Ranger: {}",
+          tenant1BucketCreatePolicy.getPolicyName());
+      auth.createAccessPolicy(tenant1BucketCreatePolicy);
+      policiesCreated.add(tenant1BucketCreatePolicy.getPolicyName());
+    } catch (Exception e) {
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  public void cleanupPolicies() {
+    for (String name : policiesCreated) {
+      try {
+        LOG.info("Deleting policy: {}", name);
+        auth.deletePolicyByName(name);
+      } catch (Exception e) {
+        LOG.error(e.getMessage());
+      }
+    }
+  }
+
+  public void cleanupRoles() {
+    for (String roleObj : rolesCreated) {
+      final JsonObject jObj = new JsonParser().parse(roleObj).getAsJsonObject();
+      final String roleId = jObj.get("id").getAsString();
+      final String roleName = jObj.get("name").getAsString();
+      try {
+        LOG.info("Deleting role: {}", roleName);
+        auth.deleteRole(roleId);
+      } catch (Exception e) {
+        LOG.error(e.getMessage());
+      }
+    }
+  }
+
+  public void cleanupUsers() {
+    for (BasicUserPrincipal user : usersCreated) {
+      try {
+        LOG.info("Deleting user: {}", user);
+        String userId = auth.getUserId(user);
+        auth.deleteUser(userId);
+      } catch (Exception e) {
+        LOG.error(e.getMessage());
+      }
+    }
+  }
+
+  public void cleanupOMDB() {
+    try {
+      omMetadataManager.getTenantStateTable().delete(TENANT_ID);
+      omMetadataManager.getTenantAccessIdTable().delete(
+          OMMultiTenantManager.getDefaultAccessId(TENANT_ID, USER_ALICE_SHORT));
+      omMetadataManager.getTenantAccessIdTable().delete(
+          OMMultiTenantManager.getDefaultAccessId(TENANT_ID, USER_BOB_SHORT));
+    } catch (IOException e) {
+      LOG.error(e.getMessage());
+    }
+  }
+
+  public void cleanupPoliciesRolesUsers() {
+    cleanupPolicies();
+    cleanupRoles();
+    cleanupUsers();
+
+    cleanupOMDB();
+  }
+
+  /**
+   * OM DB does not have the tenant state.
+   * Expect sync service to clean up all the leftover multi-tenancy
+   * policies and roles in Ranger.
+   */
+  @Test
+  public void testRemovePolicyAndRole() throws Exception {
+    long startingRangerVersion = initBGSync();
+
+    // Create roles and policies in ranger that are NOT
+    // backed up by OzoneManger Multi-Tenant tables
+    createRolesAndPoliciesInRanger(false);
+    long rangerVersionAfterCreation = bgSync.getLatestRangerServiceVersion();
+    Assert.assertTrue(rangerVersionAfterCreation >= startingRangerVersion);
+
+    bgSync.start();
+    // Wait for sync to finish once.
+    // The counter is incremented at the beginning of the run, hence the ">"
+    GenericTestUtils.waitFor(
+        () -> bgSync.getRangerSyncRunCount() > 1L, 1000, 60000);
+    bgSync.shutdown();
+    long ozoneVersionAfterSync = bgSync.getOMDBRangerServiceVersion();
+    Assert.assertTrue(ozoneVersionAfterSync > rangerVersionAfterCreation);
```

Comment:
Is this check (and other similar ones) correct? When there is divergence the order should be
1. Ozone pulls Ranger service version `x`
2. Ozone pushes updates to Ranger
3. Ozone saves version `x` to DB.

At this point we have 4 variables we should check:
`OM DB version before sync`
`OM DB version after sync`
`Ranger version before sync`
`Ranger version after sync`

I think we would expect the following for this case:
`OM DB version after sync` == `Ranger version before sync` (for the line commented here)
`Ranger version after sync` > `Ranger version before sync`
`OM DB version after sync` > `OM DB version before sync`

To ensure correctness I think all tests should have these 4 variables and compare them in the pairs shown above for each sync run using only ==, <, or > (never >= or <= which is still being used in `testRecoverRangerRole`).

Reply:
> I think we would expect the following for this case: `OM DB version after sync` == `Ranger version before sync` (for the line commented here)

I believe you mean `OM DB version after sync == Ranger version _AFTER_ sync` ?

Because `dbVerAfter == rangerVerAfter`, `rangerVerAfter == dbVerAfter > rangerVerBefore`, the latter of which is the existing check. When the DB doesn't have an initial version stored (first test in the class) it returns `-1`.

I'm adding all three checks. This should make the check logic cleaner. Thanks for the suggestion.